src/main/cpp/src/json_parser.cuh

```diff
@@ -985,59 +945,45 @@ class json_parser {
    * try skip 4 HEX chars
    * in pattern: '\\' 'u' HEX HEX HEX HEX, it's a code point of unicode
    */
-  __device__ bool try_skip_unicode(char const*& str_pos,
-                                   char const*& to_match_str_pos,
-                                   char const* const to_match_str_end,
+  __device__ bool try_skip_unicode(char_range_reader& str,
+                                   char_range_reader& to_match,
                                    char*& copy_dest,
-                                   int& unescped_string_utf8_bytes,
-                                   int& escped_string_utf8_bytes)
+                                   int& output_size_bytes)
   {
-    // already parsed u
-    bool is_success = try_skip_hex(str_pos) && try_skip_hex(str_pos) && try_skip_hex(str_pos) &&
-                      try_skip_hex(str_pos);
-    if (is_success) {
-      // parse 4 HEX chars to uint32_t value
-      auto code_point = parse_code_point(str_pos - 4);
-      auto utf_char   = codepoint_to_utf8(code_point);
-      // write utf8 bytes.
-      // In UTF-8, the maximum number of bytes used to encode a single character
-      // is 4
-      char buff[4];
-      cudf::size_type bytes = from_char_utf8(utf_char, buff);
-      unescped_string_utf8_bytes += bytes;
-      escped_string_utf8_bytes += bytes;
-
-      if (nullptr != copy_dest) {
-        for (cudf::size_type i = 0; i < bytes; i++) {
-          *copy_dest++ = buff[i];
-        }
+    // already parsed \u
+    // now we expect 4 hex chars.
+    cudf::char_utf8 code_point = 0;
+    for (size_t i = 0; i < 4; i++) {
+      if (str.eof()) { return false; }
+      char c = str.current_char();
+      str.next();
+      if (!is_hex_digit(c)) { return false; }
+      code_point = (code_point * 16) + hex_value(c);
+    }
+    auto utf_char = codepoint_to_utf8(code_point);
+    // write utf8 bytes.
+    // In UTF-8, the maximum number of bytes used to encode a single character
+    // is 4
+    char buff[4];
+    cudf::size_type bytes = from_char_utf8(utf_char, buff);
+    output_size_bytes += bytes;
+
+    // TODO I think if we do an escape sequence for \n/etc it will return
+    // the wrong thing....
+    if (nullptr != copy_dest) {
+      for (cudf::size_type i = 0; i < bytes; i++) {
+        *copy_dest++ = buff[i];
       }
+    }
 
-      if (nullptr != to_match_str_pos) {
-        for (cudf::size_type i = 0; i < bytes; i++) {
-          if (!(to_match_str_pos < to_match_str_end && *to_match_str_pos == buff[i])) {
-            return false;
-          }
-          to_match_str_pos++;
-        }
+    if (!to_match.is_null()) {
+      for (cudf::size_type i = 0; i < bytes; i++) {
+        if (!(to_match.eof() && to_match.current_char() == buff[i])) { return false; }
```

Comment:
Very mild potential improvement:  instead of checking eof() every iteration, cap the `i < bytes` by how much is left in `to_match`.  Probably not even worth the potential for creating a bug.